<a href="https://colab.research.google.com/github/Michwynn/London-Airbnb-Analysis---2/blob/Elias/Airbnb_Data_Base_Cleaning.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Configuration and library set-up**

In [1]:
# data manipulation
import pandas as pd
import numpy as np 
import re # regex
import random
from collections import Counter, defaultdict

# machine learning
from sklearn.dummy import DummyClassifier
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import f1_score
from sklearn.model_selection import train_test_split

# timeit
from tqdm import tqdm

# data visualisation
import altair as alt
import seaborn as sns
import matplotlib.pyplot as plt
from matplotlib.pyplot import figure

# set up working directory
import os
from google.colab import drive
drive.mount('/content/drive')
%cd /content/drive/My Drive/Airbnb_Milestone2

# supress warnings
import warnings 
warnings.filterwarnings('ignore')

# Display all columns
pd.set_option('display.max_columns', None)

Mounted at /content/drive
/content/drive/.shortcut-targets-by-id/1wUOfFY-ki2nFzneeaTtXLEeMjaSdKrrj/Airbnb_Milestone2


In [78]:
# read data
reviews_df = pd.read_csv('Datasets/reviews.csv') 
# remove white spaces in column headings
reviews_df.columns = reviews_df.columns.str.strip() 
# Examine dataset
display(reviews_df.head(10))
display(reviews_df.shape)

listing_id                  id        date  reviewer_id  \
0            52228441  623723762668719111  2022-05-10     37052865   
1            52228441  505671819125096360  2021-11-28     70830110   
2            52228441  466510411892882382  2021-10-05     83617224   
3            52228441  604109461995958546  2022-04-13      2152541   
4  605617198416835367  633128504578904919  2022-05-23     45418187   
5            52228441  488275073602903479  2021-11-04     37185586   
6            17501535  668631693343526023  2022-07-11      5132271   
7            52228441  588257361951713358  2022-03-22    248672143   
8            17501535  633193507110858827  2022-05-23    413544004   
9            52228441  491866989310660615  2021-11-09      8452946   

  reviewer_name                                           comments  
0      Kimberly  Great location, and the host was very responsi...  
1       Mahelet  Duccio is a lovely and friendly host. From arr...  
2          Will  Duccio is a good communicator… he was very hel...  
3     Francesco  Not entirely compliant to the pics.<br/>Good l...  
4        Waddah                         Great place and great host  
5       Ruslans  The flat is very well connected to the city by...  
6          Marc                                         Nice place  
7      Sveinung  Big apartment but a little bit run down.  Kitc...  
8        Daniel  Amazing place, we loved it. Tons of space, nic...  
9        Johnny  Duccio was a great host, super friendly with l...

(1216212, 6)

# **Data Cleaning**

**Steps to follow**

1.   Drop rows containing NAN values (DONE)
2.   Drop rows with empty string for 'comments' column (DONE)
3.   Remove HTML tags from string in 'comments' column (DONE)
4.   Remove special characters including emojis from strings (DONE)
5.   Drop rows containing non-english string in 'comments' column (DONE)
6.   Remove stopwords from string in 'comments' column
7.   Remove any short-word form string in 'comments' column






In [41]:
# check if there is any null value 
reviews_df.isnull().sum()

listing_id        0
id                0
date              0
reviewer_id       0
reviewer_name     2
comments         92
dtype: int64

In [42]:
# check if there is any empty string in the comments column
reviews_df.loc[reviews_df["comments"] == "."]

listing_id                  id        date  reviewer_id  \
92         21526298  675869744229404323  2022-07-21    215666012   
319        32537244           572622781  2019-12-03    310625523   
2388        3370968  551274991374749641  2022-01-30     88466671   
3143       13396482           562849095  2019-11-11    299309771   
3334          41509            49170962  2015-10-01       736090   
...             ...                 ...         ...          ...   
1209963    31495400           518304382  2019-08-26     34520842   
1212428    31201633           419606016  2019-03-04    243989008   
1212852    31304270           419096711  2019-03-03    244943897   
1214589    32002019           474417589  2019-06-23     64631352   
1214718    32008953           427250653  2019-03-23     42680816   

        reviewer_name comments  
92              David        .  
319          Fernando        .  
2388            Bekee        .  
3143          Cecilie        .  
3334          Alberto        .  
...               ...      ...  
1209963       Augusto        .  
1212428        Javier        .  
1212852       Bernard        .  
1214589      Margaret        .  
1214718     Christina        .  

[1477 rows x 6 columns]

In [79]:
# Drop the null value and comments value with "."
reviews_df = reviews_df[reviews_df['comments'].notna()]
display(reviews_df.shape)

(1216120, 6)

---
92 Comments containing null value dropped

In [80]:
# Drop the rows with empty string for comments column
reviews_df = reviews_df.loc[reviews_df["comments"] != "."]
display(reviews_df.shape)

(1214643, 6)

---
1477 Comments containing empty string dropped

In [91]:
# helper function to remove HTML tags from comments
from bs4 import BeautifulSoup
copy_df = reviews_df.copy()
#small_df = copy_df.iloc[:100,]
copy_df['comments'] = copy_df['comments'].apply(lambda s: BeautifulSoup(s).text)
display(copy_df.head(100))

listing_id                  id        date  reviewer_id  \
0              52228441  623723762668719111  2022-05-10     37052865   
1              52228441  505671819125096360  2021-11-28     70830110   
2              52228441  466510411892882382  2021-10-05     83617224   
3              52228441  604109461995958546  2022-04-13      2152541   
4    605617198416835367  633128504578904919  2022-05-23     45418187   
..                  ...                 ...         ...          ...   
96             17501535  588209695468203799  2022-03-22     31576937   
97             32537244           617521403  2020-03-13     84592669   
98             17501535           594341871  2020-01-18     84447796   
99             17501535  561415884152066004  2022-02-13    147957618   
100            32537244           615635811  2020-03-08     20813991   

    reviewer_name                                           comments  
0        Kimberly  Great location, and the host was very responsi...  
1         Mahelet  Duccio is a lovely and friendly host. From arr...  
2            Will  Duccio is a good communicator… he was very hel...  
3       Francesco  Not entirely compliant to the pics.Good locati...  
4          Waddah                         Great place and great host  
..            ...                                                ...  
96         Julian  Superb hospitality. Dominic was friendly, resp...  
97        Bun-Hok  Good place to stay, although a bit noisy, as i...  
98        Mallory  Easy and convenient check-in/check-out.  The s...  
99         Sandra  Dominic was a great host! We had to reach out ...  
100      Munteanu  Visited London over the weekend together with ...  

[100 rows x 6 columns]

In [92]:
# removing special characters including punctuations and emojis. 
copy2_df = copy_df.copy()

copy2_df['comments'] = copy2_df['comments'].str.replace('[^\w\s#@/:%.,_-]', '', flags=re.UNICODE)
copy2_df.columns = copy2_df.columns.str.strip() 

display(copy2_df)

listing_id                  id        date  reviewer_id  \
0                  52228441  623723762668719111  2022-05-10     37052865   
1                  52228441  505671819125096360  2021-11-28     70830110   
2                  52228441  466510411892882382  2021-10-05     83617224   
3                  52228441  604109461995958546  2022-04-13      2152541   
4        605617198416835367  633128504578904919  2022-05-23     45418187   
...                     ...                 ...         ...          ...   
1216207            14832630           184884203  2017-08-20     54407484   
1216208            14832630           173531056  2017-07-23    130284505   
1216209            14832630           153170957  2017-05-20    119296298   
1216210            14832630           145456398  2017-04-18     16394435   
1216211            14832630           103611593  2016-09-23     65608870   

        reviewer_name                                           comments  
0            Kimberly  Great location, and the host was very responsi...  
1             Mahelet  Duccio is a lovely and friendly host. From arr...  
2                Will  Duccio is a good communicator he was very help...  
3           Francesco  Not entirely compliant to the pics.Good locati...  
4              Waddah                         Great place and great host  
...               ...                                                ...  
1216207            丽云  干净舒适适合家庭入住地段优越景点全部很多都可以步行到达节省了交通费房间设施齐备能满足我们的一...  
1216208     Véronique  Excellente situation au plein centre de Londre...  
1216209        Nicole  Todo perfecto  Ubicación, comodidad y Rachel c...  
1216210    Christophe  Tout dabord même si nous navons pas eu le plai...  
1216211          Ping                                          Rachel人很好  

[1214643 rows x 6 columns]

# **Comments Language Identification**

In [12]:
# installing and Importing fasttext library for language identification
!pip install fasttext
import fasttext
# settting up the pretrained model
fasttxt_model = fasttext.load_model("lid.176.bin")

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/


In [95]:
# test model with example text
fasttxt_model.predict('이 집은 정말 좋아요! 맘에 쏙 드네요')[0][0][-2:]

'ko'

In [93]:
# Adding an extra column for language identification
copy2_df['lang'] = copy2_df['comments'].apply(lambda x: fasttxt_model.predict(x)[0][0][-2:])
display(copy2_df)

listing_id                  id        date  reviewer_id  \
0                  52228441  623723762668719111  2022-05-10     37052865   
1                  52228441  505671819125096360  2021-11-28     70830110   
2                  52228441  466510411892882382  2021-10-05     83617224   
3                  52228441  604109461995958546  2022-04-13      2152541   
4        605617198416835367  633128504578904919  2022-05-23     45418187   
...                     ...                 ...         ...          ...   
1216207            14832630           184884203  2017-08-20     54407484   
1216208            14832630           173531056  2017-07-23    130284505   
1216209            14832630           153170957  2017-05-20    119296298   
1216210            14832630           145456398  2017-04-18     16394435   
1216211            14832630           103611593  2016-09-23     65608870   

        reviewer_name                                           comments lang  
0            Kimberly  Great location, and the host was very responsi...   en  
1             Mahelet  Duccio is a lovely and friendly host. From arr...   en  
2                Will  Duccio is a good communicator he was very help...   en  
3           Francesco  Not entirely compliant to the pics.Good locati...   en  
4              Waddah                         Great place and great host   en  
...               ...                                                ...  ...  
1216207            丽云  干净舒适适合家庭入住地段优越景点全部很多都可以步行到达节省了交通费房间设施齐备能满足我们的一...   en  
1216208     Véronique  Excellente situation au plein centre de Londre...   fr  
1216209        Nicole  Todo perfecto  Ubicación, comodidad y Rachel c...   es  
1216210    Christophe  Tout dabord même si nous navons pas eu le plai...   fr  
1216211          Ping                                          Rachel人很好   zh  

[1214643 rows x 7 columns]

In [116]:
# calculating number of rows with non-english comments 
display(copy2_df.lang.value_counts())
prop = (1214643-1089747)/1214643*100
f'The total non-english comments proportion is {prop}.'

en    1089747
fr      45182
es      22556
de      16384
it      10215
       ...   
br          1
mk          1
ne          1
io          1
ms          1
Name: lang, Length: 77, dtype: int64

'The total non-english comments proportion is 10.282527458685392.'

---
**The total non-english comments proportion is 10.28%. 
Therefore, it is safe to drop the rows with non-english comments.**




In [117]:
# keeping only rows with English comments
copy3_df = copy2_df[copy2_df['lang']=='en']
display(copy3_df)

listing_id                  id        date  reviewer_id  \
0                  52228441  623723762668719111  2022-05-10     37052865   
1                  52228441  505671819125096360  2021-11-28     70830110   
2                  52228441  466510411892882382  2021-10-05     83617224   
3                  52228441  604109461995958546  2022-04-13      2152541   
4        605617198416835367  633128504578904919  2022-05-23     45418187   
...                     ...                 ...         ...          ...   
1216194            14832630           107089577  2016-10-09     46576370   
1216195            14832630           105494942  2016-10-02     31582062   
1216196            14832630           104922701  2016-09-29     95543679   
1216197            14832630           102192086  2016-09-17     16267738   
1216207            14832630           184884203  2017-08-20     54407484   

        reviewer_name                                           comments lang  
0            Kimberly  Great location, and the host was very responsi...   en  
1             Mahelet  Duccio is a lovely and friendly host. From arr...   en  
2                Will  Duccio is a good communicator he was very help...   en  
3           Francesco  Not entirely compliant to the pics.Good locati...   en  
4              Waddah                         Great place and great host   en  
...               ...                                                ...  ...  
1216194        Anders  This was an excellent apartment. Very clean, w...   en  
1216195         Magda  The place was a dream come true\r Its newly re...   en  
1216196           Joe  Her place is so greatNice and clean. My wife a...   en  
1216197          Neta  Right at the heart of Soho.The place is comfor...   en  
1216207            丽云  干净舒适适合家庭入住地段优越景点全部很多都可以步行到达节省了交通费房间设施齐备能满足我们的一...   en  

[1089747 rows x 7 columns]

In [118]:
# Special case that has both chinese and English
display(copy2_df.loc[[1216207]])

listing_id         id        date  reviewer_id reviewer_name  \
1216207    14832630  184884203  2017-08-20     54407484            丽云   

                                                  comments lang  
1216207  干净舒适适合家庭入住地段优越景点全部很多都可以步行到达节省了交通费房间设施齐备能满足我们的一...   en